odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+        """
+
+        if not isinstance(convex_conj_f, Functional):
+            raise TypeError('input {} is not a Functional instance'
+                            ''.format(convex_conj_f))
+
+        if not isinstance(y, LinearSpaceVector):
+            raise TypeError('vector {!r} not a LinearSpaceVector instance.'
+                            ''.format(y))
+
+        if y not in convex_conj_f.domain:
+            raise TypeError('vector {!r} not in the domain of the functional '
+                            '{!r}.'.format(y, convex_conj_f.domain))
+
+        super().__init__(domain=convex_conj_f.domain,
+                         linear=convex_conj_f.is_linear)
+
+        self._orig_convex_conj_f = convex_conj_f
+        self._y = y
+
+    @property
+    def orig_convex_conj_f(self):
+        return self._orig_convex_conj_f
+
+    @property
+    def y(self):
+        return self._y
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+            The point in which the functional is evaluated
+
+        Returns
+        -------
+        `self(x)` : `float`
+            Evaluation of the functional.
+        """
+        return self.orig_convex_conj_f(x - self._y)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return (self.orig_convex_conj_f.gradient *
+                ResidualOperator(IdentityOperator(self.domain), self.y))
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the ConvexConjugateLinearPerturb
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+        return proximal_translation(self.orig_convex_conj_f.proximal,
+                                    self.y)(sigma)
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional.
+
+        Notes
+        -----
+        By the Fenchel-Moreau theorem, the convex conjugate of a convex
+        conjugate is the functional itself, if and only if the functional is
+        proper, convex, and lower semi-continuous.
+
+        Wikipedia article on the `Fenchel-Moreau theorem
+        <https://en.wikipedia.org/wiki/Fenchel%E2%80%93Moreau_theorem>`_.
+        """
+
+        return self.orig_convex_conj_f.conjugate_functional + self.y.T
```

Comment:
It probably is. And in the same way we should then have that the convex conjugate of `ConvexConjugateTranslation` is `ConvexConjugateLinearPerturb`, because I think the two are duals of each other in some sense.